addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/discovery/ThingDiscoveryService.java

```diff
@@ -0,0 +1,120 @@
+
+/**
+ * Every bridge will add its discovered sensors to this discovery service to make them
+ * available to the framework.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public class ThingDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+    private final List<DiscoveryResult> results = new ArrayList<>();
+    private @NonNullByDefault({}) DeconzBridgeHandler handler;
+
+    public ThingDiscoveryService() {
+        super(Stream
+                .of(BindingConstants.THING_TYPE_PRESENCE_SENSOR, BindingConstants.THING_TYPE_POWER_SENSOR,
+                        BindingConstants.THING_TYPE_DAYLIGHT_SENSOR, BindingConstants.THING_TYPE_SWITCH)
+                .collect(Collectors.toSet()), 0, true);
+    }
+
+    /**
+     * Perform a new bridge full state request.
+     */
+    @Override
+    protected void startScan() {
+        handler.requestFullState();
+    }
+
+    /**
+     * Add a sensor device to the discovery inbox.
+     *
+     * @param sensor The sensor description
+     * @param bridgeUID The bridge UID
+     */
+    private void addDevice(String sensorID, SensorMessage sensor) {
+        ThingTypeUID thingTypeUID;
+        if (sensor.type.contains("Daylight")) { // Deconz specific: Software simulated daylight sensor
+            thingTypeUID = BindingConstants.THING_TYPE_DAYLIGHT_SENSOR;
+        } else if (sensor.type.contains("Power")) { // ZHAPower, CLIPPower
+            thingTypeUID = BindingConstants.THING_TYPE_POWER_SENSOR;
+        } else if (sensor.type.contains("Presence")) { // ZHAPresence, CLIPPrensence
+            thingTypeUID = BindingConstants.THING_TYPE_PRESENCE_SENSOR;
+        } else if (sensor.type.contains("Switch")) { // ZHASwitch
+            thingTypeUID = BindingConstants.THING_TYPE_SWITCH;
+        } else if (sensor.type.contains("LightLevel")) { // ZHALightLevel
+            thingTypeUID = BindingConstants.THING_TYPE_LIGHT_SENSOR;
+        } else if (sensor.type.contains("ZHATemperature")) { // ZHATemperature
+            thingTypeUID = BindingConstants.THING_TYPE_TEMPERATURE_SENSOR;
+        } else {
+            return;
+        }
+
+        ThingUID uid = new ThingUID(thingTypeUID, handler.getThing().getUID(),
+                sensor.uniqueid.replaceAll("[^a-z0-9\\[\\]]", ""));
+
+        DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(uid).withBridge(handler.getThing().getUID())
+                .withLabel(sensor.name + " (" + sensor.manufacturername + ")").withProperty("id", sensorID)
+                .withProperty("uid", sensor.uniqueid).withRepresentationProperty("uid").build();
+        results.add(discoveryResult);
+        thingDiscovered(discoveryResult);
+    }
+
+    @Override
+    public void setThingHandler(@Nullable ThingHandler handler) {
+        this.handler = (DeconzBridgeHandler) handler;
+        this.handler.setDiscoveryService(this);
```

Comment:
You now have a circular dependency between handler and discovery service.
Would be nicer if you move the `stateRequestFinished` to a dedicated `StateRequestFinishedListener` interface, let the discovery service implement it and have the handler only depend on that one.

Reply:
Even if I hide the discovery service behind an interface, the generated byte code doesn't change. I still have a cyclic dependency. Maybe it helps to make the discovery service reference (or interface reference) within the bridge handler a weak reference.

Reply:
If you put th such an interface in the `deconz.internal.handler` package, you would only have a pckage dependency from `deconz.internal.discovery` to `deconz.internal.handler`, but not the other way round.

Reply:
True. I thought more from the perspective of the garbage collection and two circular referencing objects.